pkg/reconciler/apis/apiexportendpointsliceurls/apiexportendpointsliceurls_reconcile.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2025 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiexportendpointsliceurls
+
+import (
+	"context"
+	"net/url"
+	"path"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog/v2"
+
+	virtualworkspacesoptions "github.com/kcp-dev/kcp/cmd/virtual-workspaces/options"
+	"github.com/kcp-dev/kcp/pkg/logging"
+	apiexportbuilder "github.com/kcp-dev/kcp/pkg/virtual/apiexport/builder"
+	apisv1alpha1 "github.com/kcp-dev/kcp/sdk/apis/apis/v1alpha1"
+	corev1alpha1 "github.com/kcp-dev/kcp/sdk/apis/core/v1alpha1"
+	"github.com/kcp-dev/kcp/sdk/apis/third_party/conditions/util/conditions"
+	apisv1alpha1apply "github.com/kcp-dev/kcp/sdk/client/applyconfiguration/apis/v1alpha1"
+)
+
+type endpointsReconciler struct {
+	listShards                  func(selector labels.Selector) ([]*corev1alpha1.Shard, error)
+	getAPIExport                func(path logicalcluster.Path, name string) (*apisv1alpha1.APIExport, error)
+	listAPIBindingsByAPIExport  func(apiexport *apisv1alpha1.APIExport) ([]*apisv1alpha1.APIBinding, error)
+	patchAPIExportEndpointSlice func(ctx context.Context, cluster logicalcluster.Path, patch *apisv1alpha1apply.APIExportEndpointSliceApplyConfiguration) error
+	shardName                   string
+}
+
+type result struct {
+	url    string
+	remove bool
+}
+
+func (c *controller) reconcile(ctx context.Context, apiExportEndpointSlice *apisv1alpha1.APIExportEndpointSlice) error {
+	r := &endpointsReconciler{
+		listShards:                  c.listShards,
+		getAPIExport:                c.getAPIExport,
+		listAPIBindingsByAPIExport:  c.listAPIBindingsByAPIExport,
+		shardName:                   c.shardName,
+		patchAPIExportEndpointSlice: c.patchAPIExportEndpointSlice,
+	}
+
+	return r.reconcile(ctx, apiExportEndpointSlice)
+}
+
+func (r *endpointsReconciler) reconcile(ctx context.Context, apiExportEndpointSlice *apisv1alpha1.APIExportEndpointSlice) error {
+	if !conditions.IsTrue(apiExportEndpointSlice, apisv1alpha1.APIExportEndpointSliceReadyForURLs) {
+		return nil
+	}
+
+	s := apiExportEndpointSlice.Status.ShardSelector
+	if s == "" { // should never happen.
+		return nil
+	}
+
+	selector, err := labels.Parse(s)
+	if err != nil {
+		return err
```

Comment:
err means just requeue, right? That's not helping, is it? The string won't change.